src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -617,112 +617,321 @@ namespace smt::noodler {
             case ConversionType::TO_INT:
             case ConversionType::FROM_INT:
-                //???
-                util::throw_error("unimplemented");
+            {
+                // s = string_var, i = int_var
+
+                // s = s_1 ... s_n, subst_vars = <s_1, ..., s_n>
+                const std::vector<BasicTerm>& subst_vars = solution.get_substituted_vars(string_var);
+
+                // cases should be the collection of all words w = w_1 ... w_n, where w_i is the word of the language L_i of the automaton for s_i
+                std::vector<std::vector<mata::Word>> cases = {{}};
+                for (const BasicTerm& subst_var : solution.get_substituted_vars(string_var)) {
+                    // TODO split automaton to only-digit and some-non-digit part and do something smarter with some-non-digit part
+                    auto aut = solution.aut_ass.at(subst_var);
+                    if (!aut->is_acyclic()) {
+                        STRACE("str-conversion", tout << "failing NFA:" << *aut << std::endl;);
+                        util::throw_error("cannot process to_int/from_int for automaton with infinite language");
+                    }
+
+                    std::vector<std::vector<mata::Word>> new_cases;
+                    for (auto word : aut->get_words(aut->num_of_states())) {
+                        for (const auto& old_case : cases) {
+                            std::vector<mata::Word> new_case = old_case;
+                            new_case.push_back(word);
+                            new_cases.push_back(new_case);
+                        }
+                    }
+                    cases = new_cases;
+                }
+
+                LenNode cases_as_formula(LenFormulaType::OR);
+
+                for (const auto& one_case : cases) {
+                    assert(subst_vars.size() == one_case.size());
+
+                    mata::Word full_word; // the word w
+                    LenNode formula_for_case(LenFormulaType::AND); // conjunct C
+                    for (unsigned i = 0; i < subst_vars.size(); ++i) {
+                        const BasicTerm& subst_var = subst_vars[i]; // var s_i
+                        mata::Word word_of_subst_var = one_case[i]; // word w_i
+
+                        // creating formula
+                        //   |s_i| == |w_i| && s_i!to_int = to_int('1'.w_i) [ && s_i!to_code = to_code(w_i) ]
+
+                        // |s_i| = |w_i|
+                        formula_for_case.succ.emplace_back(LenFormulaType::EQ, std::vector<LenNode>{ subst_var, word_of_subst_var.size() });
+
+                        // s_i!to_int = to_int('1'.w_i)
+                        // we add 1 at the beginning for the cases with 0s at the beginning of w_i,
+                        // so that for example if w_i="00013", it does not turn it into 13 but into 100013
+                        formula_for_case.succ.push_back(word_to_int(word_of_subst_var, to_int_var(subst_var), true, false));
```

Comment:
I don't see what is it good for.

Reply:
We need to keep exact values in `s_i!to_int` for synchronization, so that if we have "00013" and "013", they are kept as different values (basically we encode the string into int variable).
For example, if we have `i = to_int(s)`, `i' = to_int(s')` and `s = s1 s2`, `s' = s2 s3`, it is possible that`i = to_int(s)` would be satisfiable if only `s2 = "00013"` but `i' = to_int(s')` is satisfiable if only `s2 = "013"`, so this should be unsatisfiable - we need to differentiate the cases when `s2 = "00013"` and when `s2 = "013"`.